samples/switchless/README.md

```diff
@@ -0,0 +1,184 @@
+- Recommend the number of host worker threads required for switchless calls in practice
+
+Prerequisite: you may want to read [Common Sample Information](../README.md#common-sample-information) before going further.
+
+## Switchless Calls
+
+In an enclave application, the host makes **ECALL**s into functions exposed by the enclaves it created. Likewise,
+the enclaves may make **OCALL**s into functions exposed by the host that created them. In either case, the
+execution has to be transitioned from an untrusted environment to a trusted environment, or vice versa. Since the
+transition is costly due to heavy security checks, it might be more performance advantageous to make the calls
+**context-switchless**: the caller delegates the function call to a worker thread in the other environment, which does the real job of calling the function and post the result to the caller. Both the calling thread and the
+worker thread never leave their respective execution contexts during the perceived function call.
+
+The calling thread and the worker thread need to exchange information twice during the call. When the switchless
+call is initiated, the caller needs to pass the `job` (encapsulating information regarding the function call in a
+ single object, for details see the next section) to the worker thread. And when the call finishes, the worker
+thread needs to pass the result back to the caller. Both exchanges need to be synchronized.
+
+While switchless calls save transition time, they require at least one additional thread to service the calls.
+More threads typically means more competition of the CPU cores and more context switches, hurting the performance.
+Whether to make a particular function switchless has to weigh the associated costs and savings. In general, the
+good candidates for switchless calls are functions that are: 1) short, thus the transition takes relatively high
+percentage of the overall execution time of the call; and 2) called frequently, so the savings in transition time
+add up.
+
+## How does OpenEnclave support switchless OCALLs
+
+OpenEnclave only supports synchronous switchless OCALLs currently. When the caller within an enclave makes a
+switchless OCALL, the trusted OpenEnclave runtime creates a `job` out of the function call. The `job` object
+includes information such as the function ID, the parameters marshaled into a buffer, and a buffer for holding the
+return value(s). The job is posted to a shared memory region which both the enclave and the host can access.
+
+A host worker thread checks and retrieves `job` from the shared memory region. It uses the untrusted OpenEnclave
+runtime to process the `job` by unmarshaling the parameters, then dispatching to the callee function, and finally
+relaying the result back to the trusted OpenEnclave runtime, which is further forwarded back to the caller.
+
+To support simultaneous switchless OCALLs made from enclaves, the host workers are multi-threaded. OpenEnclave
```

Comment:
"To support simultaneous switchless OCALLs made from enclaves" only applies if there can be multiple simultaneous ECALLs, which not all TEEs or enclaves support.   Maybe prefix this with something like "If an enclave supports multiple simultaneous ECALLs, each ECALL might attempt an OCALL, resulting in simultaneous OCALLs."

Reply:
Thanks for the suggestion.